### PR TITLE
[UIE-51] Add useUniqueId to components package

### PR DESCRIPTION
**Pull request by nawatts**
https://broadworkbench.atlassian.net/browse/UIE-51

Moving dependencies of Clickable to the components package... This moves `useUniqueId`.

It makes a few changes:
- Use `useState` instead of `useMemo` to guarantee the stable IDs promised in the docstring.
- Update the example with the most likely usage (IDs for linking elements).
- Simplify the test to not mock Lodash.

Alternatively, we could use React's built-in [`useId`](https://react.dev/reference/react/useId). However, it's only in React 18+, so using our own hook would allow us to more easily support React 17 in components if we choose to.

### CODE REVIEW

**Review by cahrens — APPROVED** [review]

File: packages/components/src/hooks/useUniqueId.test.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    // Assert
+    expect(resultAfterSecondRender).toEqual(resultAfterFirstRender);
+  });
+
+  it('returns a string that starts with the given prefix', () => {
+    // Act
+    const { result: hookReturnRef } = renderHook(useUniqueId, { initialProps: 'test-prefix' });
+
+    // Assert
```
Comment (dmohs) [inline packages/components/src/hooks/useUniqueId.test.ts:31]: I'm still not convinced these comments pay for themselves, but, more generally, this is a really nice test.

**Review by dmohs — APPROVED** [review]

File: packages/components/src/hooks/useUniqueId.test.ts (hunk 1 above)
Comment (nawatts) [reply packages/components/src/hooks/useUniqueId.test.ts:31]: Especially in simple tests, I agree that they don't have much descriptive value for a reader.

I find their value to be in writing tests. Forming the habit of always writing these comments serves as a useful reminder on how to structure the test.

Comment (msilva-broad) [reply packages/components/src/hooks/useUniqueId.test.ts:31]: yup, habit and test structure is a key motivation.

**Review by msilva-broad — APPROVED** [review]
nice small PR.  looks great!

**Comment by msilva-broad** [comment]
oops, forgot package version bump

**Review by msilva-broad — DISMISSED** [review]
add package version bump

**Comment by nawatts** [comment]
See reply on https://github.com/DataBiosphere/terra-ui/pull/4234#discussion_r1327468186.

I have multiple PRs open at once for components. I don't want to bump the version in every one and deal with rebasing. I'll bump the version once Clickable is in and a new version becomes useful to AoU.

**Comment by nawatts** [comment]
Rebased to resolve merge conflict in packages/src/index.ts after merging #4234.

**Pull request merged**